scanner/metadata/cuesheet/cuesheet.go

```diff
@@ -0,0 +1,675 @@
+type RemData map[string]string
+
+const (
+	// None - no flags
+	None   Flags = iota
+	Dcp          = 1 << iota
+	FourCh       = 1 << iota
+	Pre          = 1 << iota
+	Scms         = 1 << iota
+)
+
+var (
+	ErrorFrameFormat = fmt.Errorf("invalid frame format")
+)
+
+// TrackIndex data
+type TrackIndex struct {
+	Number uint
+	Frame  Frame
+}
+
+func frameFromString(str string) (Frame, error) {
+	v := strings.Split(str, ":")
+	if len(str) < 8 {
+		return 0, ErrorFrameFormat
+	}
+	if len(v) == 3 {
+		mm, _ := strconv.ParseUint(v[0], 10, 32)
+		ss, _ := strconv.ParseUint(v[1], 10, 32)
+		if ss > 59 {
+			return 0, ErrorFrameFormat
+		}
+		ff, _ := strconv.ParseUint(v[2], 10, 32)
+		if ff >= framesPerSecond {
+			return 0, ErrorFrameFormat
+		}
+		return Frame((mm*60+ss)*framesPerSecond + ff), nil
+	}
+	return 0, ErrorFrameFormat
+}
+
+func (frame Frame) Duration() time.Duration {
+	seconds := time.Second * time.Duration(frame/framesPerSecond)
+	milliseconds := time.Millisecond * time.Duration(float64(frame%framesPerSecond)/0.075)
+	return seconds + milliseconds
+}
+
+func (frame Frame) String() string {
+	seconds := frame / framesPerSecond
+	minutes := seconds / 60
+	seconds %= 60
+	ff := frame % framesPerSecond
+	return fmt.Sprintf("%.2d:%.2d:%.2d", minutes, seconds, ff)
+}
+
+// Track instance
+type Track struct {
+	Rem           RemData
+	TrackNumber   uint
+	TrackDataType string
+	Flags         Flags
+	ISRC          string
+	Title         string
+	Performer     string
+	SongWriter    string
+	PreGap        Frame
+	PostGap       Frame
+	Index         []TrackIndex
+}
+
+// File instance
+type File struct {
+	FileName string
+	FileType string
+	Tracks   []Track
+}
+
+// Cuesheet instance
+type Cuesheet struct {
+	Rem        RemData
+	Catalog    string
+	CdTextFile string
+	Title      string
+	Performer  string
+	SongWriter string
+	Pregap     Frame
+	Postgap    Frame
+	File       []File
+}
+
+func (track *Track) GetStartOffset() time.Duration {
+	if len(track.Index) > 1 {
+		return track.Index[1].Frame.Duration()
+	}
+	return track.Index[0].Frame.Duration()
+}
+
+// ReadFromFile load CUESheet from file
+func ReadFromFile(filePath string) (*Cuesheet, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := file.Close(); err != nil {
+			log.Error("Can't close CUE file", "filePath", filePath)
+		}
+	}()
+	return ReadCue(file)
+}
+
+// ReadCue loads and parses CUESHEET from reader
+func ReadCue(r io.Reader) (*Cuesheet, error) {
+	b := bufio.NewReader(r)
+	cuesheet := &Cuesheet{}
+
+	bom, _, err := b.ReadRune()
+
+	if err != nil {
+		return nil, err
+	}
+
+	if bom != '\uFEFF' {
+		if err := b.UnreadRune(); err != nil {
+			return nil, err
+		}
+	}
+
+	for {
+		line, err := (*b).ReadString('\n')
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		line = strings.Trim(line, delims)
+		command := readString(&line)
+
+		switch command {
```

Comment:
Should we add logic to extract the `FLAGS` command as well, even if it's not going to be used yet? One (admittedly low-priority) feature I think it would be nice for Navidrome to have is to be able to de-emphasize tracks that have CD pre-emphasis (and maybe I'd get around to working on a PR). So knowing which cue tracks have `FLAGS PRE` set would be useful in that case. Of course, the cue extractor could always be updated at that time.

Reply:
Nevermind, I see you already handled it elsewhere!